### PR TITLE
Add `--dont-require-trivial-properties` to ignore trivial getters and setters

### LINKED ISSUE #143: Exclude checks on simple property accessors (getters and setters)
I'm looking at adding javadoc checks to a project and this looks like a great option!  

I would like to skip this requirement for trivial javadocs on getters and setters.  The following javadocs are completely useless:
```java
/**
 * Gets the foo.
 *
 * @return the foo
 */
public Foo getFoo() {
    return this.foo;
}

/**
 * Sets the foo.
 *
 * @param foo the foo to set
 */
public void setFoo(Foo foo) {
    this.foo = foo;
}
```
Some IDE linters have this as an option in their checks, e.g., IntelliJ IDEA's "Declaration has Javadoc problems" has a checkbox to "Ignore simple Property Accessors".

Checkstyle (which I agree is heavyweight) has an option for [MissingJavadocMethodCheck](https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocMethodCheck.html) to `allowMissingPropertyJavadoc` if it matches the format above.

### PR DESCRIPTION
**Pull request by mernst**
Fixes #143

### CODE REVIEW

File: src/main/java/org/plumelib/javadoc/RequireJavadoc.java [hunk 1]
```diff
@@ -289,6 +320,158 @@ boolean shouldExclude(Path path) {
...
+      return null;
+    }
+    String upperCaseProperty = methodName.substring(3);
+    if (upperCaseProperty.length() == 0) {
+      return null;
+    }
+    if (!Character.isUpperCase(upperCaseProperty.charAt(0))) {
+      return null;
+    }
+    String lowerCaseProperty =
```
Comment (dbwiddis) [inline src/main/java/org/plumelib/javadoc/RequireJavadoc.java:429]: Similar to above, `lowerCamelCaseProperty` or if you want to have fun, `dromedaryCaseProperty`.

File: src/main/java/org/plumelib/javadoc/RequireJavadoc.java [hunk 2]
```diff
@@ -289,6 +320,158 @@ boolean shouldExclude(Path path) {
...
+   * Returns the lower-cased name of the property, if the method is a getter or setter. Otherwise
+   * returns null.
+   *
+   * @param md the method to test
+   * @param isGetter true if testing for a getter, false if testing for a setter
+   * @return the lower-cased names of the property, or null
+   */
+  @Nullable String propertyName(MethodDeclaration md, boolean isGetter) {
+    String methodName = md.getNameAsString();
+    if (!methodName.startsWith(isGetter ? "get" : "set")) {
```
[the remark this answers is not in the document]
Comment (mernst) [reply src/main/java/org/plumelib/javadoc/RequireJavadoc.java:419]: That is handled by the `if (upperCaseProperty.length() == 0) {` test a few lines below.

There are no `IndexOutOfBounds` exceptions in this code (at least for arrays and Strings) because each commit is [formally verified](https://github.com/plume-lib/require-javadoc/blob/master/build.gradle#L67) to [use correct indexing](https://checkerframework.org/manual/#index-checker).

I have added a test case, though; thanks for pointing out that this corner case was not tested.

File: src/main/java/org/plumelib/javadoc/RequireJavadoc.java [hunk 3]
```diff
@@ -289,6 +320,158 @@ boolean shouldExclude(Path path) {
...
+      return null;
+    }
+    String upperCaseProperty = methodName.substring(3);
```
[the remark this answers is not in the document]
Comment (mernst) [reply src/main/java/org/plumelib/javadoc/RequireJavadoc.java:422]: Done.  Thanks for the suggestion.

I don't think many people remember programming in Pascal -- I haven't done so since the mid-80s.  :-)

[the remark this answers is not in the document]
Comment (dbwiddis) [reply src/main/java/org/plumelib/javadoc/RequireJavadoc.java:422]: > I don't think many people remember programming in Pascal -- I haven't done so since the mid-80s. :-)

Wrote my master's thesis code with it in the mid-90s! 😁

**Comment by mernst** [comment]
@dbwiddis Please take another look.  The code now handles `isX()` getters, and some of the code you didn't like is cleaner via use of an enum.

File: src/main/java/org/plumelib/javadoc/RequireJavadoc.java [hunk 4]
```diff
@@ -289,6 +320,194 @@ boolean shouldExclude(Path path) {
     return shouldExclude(path.toString());
   }
 
+  /** The type of property method: a getter, boolean getter, or setter. */
+  enum PropertyType {
+    /** A method of the form {@code int getFoo()}. */
+    GETTER("get", 0, false),
+    /** A method of the form {@code boolean isFoo()}. */
+    GETTER_IS("is", 0, false),
```
Comment (peternied) [inline src/main/java/org/plumelib/javadoc/RequireJavadoc.java:328]: Considering also including `has` which is sometimes more natural than `is`

File: README.md [hunk 5]
```diff
@@ -28,6 +28,7 @@ Usage: java org.plumelib.javadoc.RequireJavadoc [options] [directory-or-file ...
   --exclude=<regex>                - Don't check files or directories whose pathname matches the regex
   --dont-require=<regex>           - Don't report problems in Java elements whose name matches the regex
   --dont-require-private=<boolean> - Don't report problems in elements with private access [default: false]
+  --dont-require-properties=<boolean> - Don't report problems about trivial getters and setters [default: false]
```
Comment (peternied) [inline README.md:31]: Should the value be verbose `--dont-require-trivial-properties` so that there could be a `--dont-require-properties` use case is also available?

**Review by peternied — COMMENTED** [review]
Nice improvement!

File: src/main/java/org/plumelib/javadoc/RequireJavadoc.java [hunk 6]
```diff
@@ -289,6 +320,200 @@ boolean shouldExclude(Path path) {
...
   }
 
+  /** The type of property method: a getter or setter. */
+  enum PropertyType {
+    /** A method of the form {@code int getFoo()}. */
+    GETTER("get", 0, false),
+    /** A method of the form {@code boolean hasFoo()}. */
+    GETTER_HAS("has", 0, false),
+    /** A method of the form {@code boolean isFoo()}. */
+    GETTER_IS("is", 0, false),
```
Comment (dbwiddis) [inline src/main/java/org/plumelib/javadoc/RequireJavadoc.java:330]: I'm perfectly happy with `get`, `set`, and `is`.  However, since you introduced has, I'll just toss out there that `not` is sometimes common (the opposite of `is` for a boolean getter.)
```java
private boolean initialized = false;

public boolean notInitialized() {
  return !this.initialized;
}
```
I've also seen `unsetFoo()` to change a value to its default (or null).  Not suggesting you add that, either, just offering commentary for completeness.

**Review by peternied — APPROVED** [review]
@mernst Can't wait for version `1.0.4` :ship:

**Review by dbwiddis — APPROVED** [review]

File: src/main/java/org/plumelib/javadoc/RequireJavadoc.java (hunk 6 above)
Comment (mernst) [reply src/main/java/org/plumelib/javadoc/RequireJavadoc.java:351]: I added `not` since it was easy.  I'm not going to add `unset` -- at least, not now.

Comment (dbwiddis) [reply src/main/java/org/plumelib/javadoc/RequireJavadoc.java:351]: Agreed, as in particular, "unset" is really "set to some empty or default thing" which needs a javadoc.

**Comment by mernst** [comment]
> Can't wait for version `1.0.4`

I have released version 1.0.4.  The release is available on Maven Central and on GitHub.

**Comment by dbwiddis** [comment]
> I have released version 1.0.4. The release is available on Maven Central and on GitHub.

Thanks for the quick response and release, all within the space of a PR I started on Monday!
